File: 백준/Gold/1774. 우주신과의 교감/우주신과의 교감.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

const int MX = 1010;
int n, m;
int par[MX];
int arr[MX][2];
double res;

vector<pair<double,pair<int,int>>> vec;

int find(int x){
    if(par[x] == x) return x;
    else return par[x] = find(par[x]);
}

void union_(int a, int b){
    a = find(a);
    b = find(b);
    
    if(a == b) return;
    
    par[a] = b;
}

int main()
{
    cin >> n >> m;
    
    for(int i = 1; i <= n; i++){
        par[i] = i;
    }
    
    for(int i = 1; i <= n; i++) {
        cin >> arr[i][0] >> arr[i][1];
    }
    
    for(int i = 0; i < m;i++){
        int a, b;
        cin >> a >> b;
        
        union_(a, b);
    }
  
    for(int i = 1; i <= n; i++){
        for(int j = i+1; j <= n; j++){
            long dx = arr[i][0] - arr[j][0];
            long dy = arr[i][1] - arr[j][1];
            
            double d = sqrt(dx * dx + dy * dy);
            
            vec.push_back({d, {i,j}});
        }
    }
    
    sort(vec.begin(), vec.end());
    
    for(int i= 0; i< vec.size(); i++){
        int a = find(vec[i].second.second);
        int b = find(vec[i].second.first);
        double c = vec[i].first;
        
        if(a == b) continue;
      
        union_(a, b);

        res += c;
    }
    
    printf("%.2f",res);
    return 0;
}

File: 백준/Gold/1774. 우주신과의 교감/우주신과의 교감.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.StringTokenizer;

public class Main {
	static int par[] = new int[1020];
	static ArrayList<Vertex> list = new ArrayList<>();
	static ArrayList<Edge> vec = new ArrayList<>();

	static class Vertex {
		int id;
		double x, y;

		Vertex(int id, double x, double y) {
			this.id = id;
			this.x = x;
			this.y = y;
		}
	}

	static class Edge {
		double d;
		Vertex a, b;

		public Edge() {

		}

		public Edge(Vertex a, Vertex b) {
			this.a = a;
			this.b = b;
			this.d = getDist();
		}

		double getDist() {
			double diffX = Math.abs(a.x - b.x);
			double diffY = Math.abs(a.y - b.y);
			return Math.sqrt(Math.pow(diffX, 2) + Math.pow(diffY, 2));
		}
	}

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());

		init();
		list.add(new Vertex(-1, -1, -1));
		for (int i = 1; i <= N; i++) {
			st = new StringTokenizer(br.readLine());
			Double x = Double.parseDouble(st.nextToken());
			Double y = Double.parseDouble(st.nextToken());

			list.add(new Vertex(i, x, y));
		}

		for (int i = 0; i < M; i++) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());

			union(a, b);
		}

		for (int i = 1; i <= N; i++) {
			Vertex a = list.get(i);
			for (int j = i + 1; j <= N; j++) {
				Vertex b = list.get(j);
				vec.add(new Edge(a, b));
			}
		}

		Collections.sort(vec, (o1, o2) -> Double.compare(o1.d, o2.d));

		double tot = 0;

		for (Edge edge : vec) {
			if (find(edge.a.id) == find(edge.b.id))
				continue;
			union(edge.a.id, edge.b.id);
			tot += edge.d;
		}

		System.out.println(String.format("%.2f", tot));
	}

	static void init() {
		for (int i = 0; i < 1020; i++) {
			par[i] = i;
		}
	}

	static int find(int a) {
		if (par[a] == a)
			return a;
		return par[a] = find(par[a]);
	}

	static void union(int a, int b) {
		a = find(a);
		b = find(b);
		if (a == b)
			return;

		par[b] = a;
	}
}